File: nodeset/nodeset.go
// Copyright 2019 Grendel Authors. All rights reserved.
//
// This file is part of Grendel.
//
// Grendel is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Grendel is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Grendel. If not, see <https://www.gnu.org/licenses/>.

package nodeset

import (
	"encoding/json"
	"errors"
	"fmt"
	"regexp"
	"sort"
	"strings"
)

var (
	ErrInvalidNodeSet = errors.New("invalid nodeset")
	ErrParseNodeSet   = errors.New("nodeset parse error")
	rangeSetRegexp    = regexp.MustCompile(`(\[[^\[\]]+\]|[0-9]+)`)
)

type Pattern struct {
	format   string
	rangeSet *RangeSetND
}

type NodeSet struct {
	patterns map[string]*RangeSetND
}

func NewNodeSet(nodestr string) (*NodeSet, error) {
	ns := &NodeSet{patterns: make(map[string]*RangeSetND, 0)}
	err := ns.Add(nodestr)
	if err != nil {
		return nil, err
	}

	return ns, nil
}

func (ns *NodeSet) Add(nodestr string) error {
	nodestr = strings.ReplaceAll(nodestr, " ", "")
	if nodestr == "" {
		return fmt.Errorf("empty nodeset - %w", ErrParseNodeSet)
	}

	ranges := rangeSetRegexp.FindAllStringSubmatch(nodestr, -1)
	patterns := rangeSetRegexp.ReplaceAllString(nodestr, "%s")

	if strings.Index(patterns, "[") != -1 {
		return fmt.Errorf("unbalanced '[' found while parsing %s - %w", nodestr, ErrParseNodeSet)
	}
	if strings.Index(patterns, "]") != -1 {
		return fmt.Errorf("unbalanced ']' found while parsing %s - %w", nodestr, ErrParseNodeSet)
	}

	ridx := 0
	for _, pattern := range strings.Split(patterns, ",") {
		rangeSetCount := strings.Count(pattern, "%s")
		if rangeSetCount == 0 {
			ns.patterns[pattern] = nil
			continue
		}

		rangeSets := make([]string, 0)
		for i := ridx; i < ridx+rangeSetCount; i++ {
			rangeSets = append(rangeSets, strings.Trim(ranges[i][1], "[]"))
		}

		rs, err := NewRangeSetND([][]string{rangeSets})
		if err != nil {
			return err
		}

		if _, ok := ns.patterns[pattern]; !ok {
			ns.patterns[pattern] = rs
		} else {
			err = ns.patterns[pattern].Update(rs)
			if err != nil {
				return err
			}
		}

		ridx += rangeSetCount
	}

	return nil
}

func (ns *NodeSet) Len() int {
	size := 0

	for _, rs := range ns.patterns {
		if rs == nil {
			size++
		} else {
			size += rs.Len()
		}
	}

	return size
}

func (ns *NodeSet) String() string {
	list := ns.toStringList()
	return strings.Join(list, ",")
}

func (ns *NodeSet) toStringList() []string {
	list := make([]string, 0)

	items := make([]*Pattern, 0, len(ns.patterns))
	for pattern, rs := range ns.patterns {
		items = append(items, &Pattern{format: pattern, rangeSet: rs})
	}

	sort.SliceStable(items, func(i, j int) bool {
		return items[i].rangeSet.Len() > items[j].rangeSet.Len()
	})

	for _, pattern := range items {
		if pattern.rangeSet == nil {
			list = append(list, pattern.format)
			continue
		}

		for _, params := range pattern.rangeSet.FormatList() {
			list = append(list, fmt.Sprintf(pattern.format, params...))
		}

	}

	return list
}

func (ns *NodeSet) MarshalJSON() ([]byte, error) {
	list := ns.toStringList()
	return json.Marshal(&list)
}

func (ns *NodeSet) UnmarshalJSON(data []byte) error {
	var list []string
	if err := json.Unmarshal(data, &list); err != nil {
		return err
	}

	n, err := NewNodeSet(strings.Join(list, ","))
	if err != nil {
		return err
	}

	*ns = *n

	return nil
}

func (ns *NodeSet) Iterator() *NodeSetIterator {
	nodes := make([]string, 0, ns.Len())

	items := make([]*Pattern, 0, len(ns.patterns))
	for pattern, rs := range ns.patterns {
		items = append(items, &Pattern{format: pattern, rangeSet: rs})
	}

	sort.SliceStable(items, func(i, j int) bool {
		return items[i].rangeSet.Len() > items[j].rangeSet.Len()
	})

	for _, pattern := range items {
		if pattern.rangeSet == nil {
			nodes = append(nodes, pattern.format)
			continue
		}

		it := pattern.rangeSet.Iterator()
		for it.Next() {
			params := it.FormatList()
			nodes = append(nodes, fmt.Sprintf(pattern.format, params...))
		}

	}

	return &NodeSetIterator{nodes: nodes, current: -1}
}

File: internal/frontend/src/lib/nodeset/node-set.ts
import { ParseNodeSetError } from "./errors";
import { formatPattern } from "./format";
import { NodeSetIterator } from "./iterator";
import { RangeSetND } from "./range-set-nd";

const rangeSetRegexp = /(\[[^[\]]+\]|[0-9]+)/g;

interface PatternEntry {
  format: string;
  rangeSet: RangeSetND | null;
}

export class NodeSet implements Iterable<string> {
  private readonly patterns = new Map<string, RangeSetND | null>();

  constructor(nodestr = "") {
    if (nodestr === "") return;
    this.add(nodestr);
  }

  add(nodestr: string): void {
    const cleaned = nodestr.replace(/ /g, "");
    if (cleaned === "") {
      throw new ParseNodeSetError("empty nodeset");
    }

    const ranges: string[] = [];
    for (const m of cleaned.matchAll(rangeSetRegexp)) {
      ranges.push(m[1]);
    }
    const patterns = cleaned.replace(rangeSetRegexp, "%s");

    if (patterns.indexOf("[") !== -1) {
      throw new ParseNodeSetError(`unbalanced '[' found while parsing ${cleaned}`);
    }
    if (patterns.indexOf("]") !== -1) {
      throw new ParseNodeSetError(`unbalanced ']' found while parsing ${cleaned}`);
    }

    let ridx = 0;
    for (const pattern of patterns.split(",")) {
      const rangeSetCount = countOccurrences(pattern, "%s");
      if (rangeSetCount === 0) {
        this.patterns.set(pattern, null);
        continue;
      }

      const rangeSets: string[] = [];
      for (let i = ridx; i < ridx + rangeSetCount; i++) {
        rangeSets.push(trimBrackets(ranges[i]));
      }

      const rs = new RangeSetND([rangeSets]);
      const existing = this.patterns.get(pattern);
      if (existing == null) {
        this.patterns.set(pattern, rs);
      } else {
        existing.update(rs);
      }

      ridx += rangeSetCount;
    }
  }

  len(): number {
    let size = 0;
    for (const rs of this.patterns.values()) {
      if (rs === null) size++;
      else size += rs.len();
    }
    return size;
  }

  toString(): string {
    return this.toStringList().join(",");
  }

  toStringList(): string[] {
    const items: PatternEntry[] = [];
    for (const [format, rangeSet] of this.patterns) {
      items.push({ format, rangeSet });
    }

    sortStable(items, (a, b) => {
      const alen = a.rangeSet === null ? 1 : a.rangeSet.len();
      const blen = b.rangeSet === null ? 1 : b.rangeSet.len();
      return blen - alen;
    });

    const list: string[] = [];
    for (const entry of items) {
      if (entry.rangeSet === null) {
        list.push(entry.format);
        continue;
      }
      for (const params of entry.rangeSet.formatList()) {
        list.push(formatPattern(entry.format, params));
      }
    }
    return list;
  }

  iterator(): NodeSetIterator {
    const items: PatternEntry[] = [];
    for (const [format, rangeSet] of this.patterns) {
      items.push({ format, rangeSet });
    }

    sortStable(items, (a, b) => {
      const alen = a.rangeSet === null ? 1 : a.rangeSet.len();
      const blen = b.rangeSet === null ? 1 : b.rangeSet.len();
      return blen - alen;
    });

    const nodes: string[] = [];
    for (const entry of items) {
      if (entry.rangeSet === null) {
        nodes.push(entry.format);
        continue;
      }
      const it = entry.rangeSet.iterator();
      while (it.next()) {
        nodes.push(formatPattern(entry.format, it.formatList()));
      }
    }

    return new NodeSetIterator(nodes);
  }

  toJSON(): string[] {
    return this.toStringList();
  }

  static fromJSON(value: unknown): NodeSet {
    if (!Array.isArray(value) || !value.every((v) => typeof v === "string")) {
      throw new ParseNodeSetError("expected a JSON array of strings");
    }
    return new NodeSet(value.join(","));
  }

  [Symbol.iterator](): Iterator<string> {
    return this.iterator()[Symbol.iterator]();
  }
}

function trimBrackets(s: string): string {
  let start = 0;
  let end = s.length;
  while (start < end && (s[start] === "[" || s[start] === "]")) start++;
  while (end > start && (s[end - 1] === "[" || s[end - 1] === "]")) end--;
  return s.slice(start, end);
}

function countOccurrences(haystack: string, needle: string): number {
  if (needle === "") return 0;
  let count = 0;
  let pos = 0;
  while ((pos = haystack.indexOf(needle, pos)) !== -1) {
    count++;
    pos += needle.length;
  }
  return count;
}

function sortStable<T>(arr: T[], cmp: (a: T, b: T) => number): void {
  const indexed = arr.map((value, index) => ({ value, index }));
  indexed.sort((a, b) => {
    const c = cmp(a.value, b.value);
    return c !== 0 ? c : a.index - b.index;
  });
  for (let i = 0; i < arr.length; i++) arr[i] = indexed[i].value;
}
